### PR TITLE
Error: Property is overwritten by later spread

### LINKED ISSUE #27355: Missing excess property error with spread
**TypeScript Version:**  3.1.0-dev.201xxxxx

**Code**

```ts
function f(o: { a: number, b: number }): { a: number } {
    return { b: 0, ...o };
}
```

**Expected behavior:**

Error on `b`.

**Actual behavior:**

No error.

Broken by #26798.

**Related Issues:** #27166, #23340

### PR DESCRIPTION
**Pull request by sandersn**
Move the checking from JSX to getSpreadType so that normal objects get the error too.

Also reword the missing semicolon related error to be less silly.

Fixes #27355 (probably)

### CODE REVIEW

**Comment by sandersn** [comment]
Note that the related span didn't work out because getSpreadType doesn't have easy access to the *usage* of properties, just their declarations. For spread properties, the declaration is a really bad site for a related span. I decided that the additional complexity of determining the usage site wasn't worth it.

File: tests/baselines/reference/checkJsxChildrenProperty2.errors.txt [hunk 1]
```diff
@@ -44,8 +44,8 @@ tests/cases/conformance/jsx/file.tsx(49,6): error TS2322: Type '{ children: Elem
     
     let k0 =
         <Comp a={10} b="hi" children="Random" >
-              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2710: 'children' are specified twice. The attribute named 'children' will be overwritten.
+                            ~~~~~~~~~~~~~~~~~
+!!! error TS2735: 'children' is overwritten by a later spread.
```
Comment (RyanCavanaugh) [inline tests/baselines/reference/checkJsxChildrenProperty2.errors.txt]: Error message is inaccurate here

File: tests/baselines/reference/checkJsxChildrenProperty13.errors.txt [hunk 2]
```diff
@@ -14,8 +14,8 @@ tests/cases/conformance/jsx/file.tsx(12,30): error TS2710: 'children' are specif
         render() {
             // Error children are specified twice
             return (<InnerButton {...this.props} children="hi">
-                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2710: 'children' are specified twice. The attribute named 'children' will be overwritten.
+                                                 ~~~~~~~~~~~~~
+!!! error TS2735: 'children' is overwritten by a later spread.
```
Comment (weswigham) [inline tests/baselines/reference/checkJsxChildrenProperty13.errors.txt]: ...later? Methinks this needs a different error message, since it's not being overridden by a spread, but by the actual children tags in the AST.

File: tests/baselines/reference/objectSpreadNegative.errors.txt [hunk 3]
```diff
@@ -5,22 +5,29 @@ tests/cases/conformance/types/spread/objectSpreadNegative.ts(23,1): error TS2322
...
-tests/cases/conformance/types/spread/objectSpreadNegative.ts(38,19): error TS2698: Spread types may only be created from object types.
-tests/cases/conformance/types/spread/objectSpreadNegative.ts(43,1): error TS2349: Cannot invoke an expression whose type lacks a call signature. Type '{}' has no compatible call signatures.
-tests/cases/conformance/types/spread/objectSpreadNegative.ts(47,12): error TS2339: Property 'b' does not exist on type '{}'.
-tests/cases/conformance/types/spread/objectSpreadNegative.ts(53,9): error TS2339: Property 'm' does not exist on type '{ p: number; }'.
-tests/cases/conformance/types/spread/objectSpreadNegative.ts(58,11): error TS2339: Property 'a' does not exist on type '{}'.
-tests/cases/conformance/types/spread/objectSpreadNegative.ts(62,14): error TS2698: Spread types may only be created from object types.
-tests/cases/conformance/types/spread/objectSpreadNegative.ts(65,14): error TS2698: Spread types may only be created from object types.
+tests/cases/conformance/types/spread/objectSpreadNegative.ts(27,20): error TS2735: 'b' is overwritten by a later spread.
```
Comment (weswigham) [inline tests/baselines/reference/objectSpreadNegative.errors.txt]: I don't think it's safe to emit this error outside of `strictNullChecks` in normal circumstances (JSX children being one of the few where it is).

**Comment by weswigham** [comment]
> For spread properties, the declaration is a really bad site for a related span. I decided that the additional complexity of determining the usage site wasn't worth it.

We should have a related span that calls out _which_ following spread (just in case there are multiple) is the one which overrides the property.

**Comment by andy-ms** [comment]
False positive:
```ts
function f(obj: { x: number } | undefined): { x: number } {
    return { x: 1, ...obj };
}
```
(mentioned by @ajafff in #23340)

**Comment by sandersn** [comment]
@andy-ms Actually, the related example that you gave in that issue also fails, with `obj: { x: number | undefined }`. I have a commit that fixes these two problems, but it does so by not issuing the error for any union-distributed spread (fixes the first example) and not issuing the error for any nullable-containing property type (fixes the second example). It also reverts the JsX test changes, although that issue isn’t really fixed. 

Basically, this fix is more complicated than I hoped, so I’m going to put it on hold until I address more urgent 3.2work.

**Comment by matthargett** [comment]
3.2 has been out for a while -- any updates on this fix, or one that supersedes it? :D if there were complications, can they be described in a comment so we can understand better?

**Comment by sandersn** [comment]
#36727 is the revived version of this, which is now merged.
